### PR TITLE
`GateWithRegisters.controlled()` now returns `BloqAsCirqGate(Controlled())` instead of `cirq.Controlled` gate.

### PR DESCRIPTION
**Pull request by tanujkhattar**
This is a step towards using the new `Controlled` bloq instead of `cirq.Controlled` gate throghout the library. Eventually, we'd to update all the call sites so that `GateWithRegisters.controlled` follows the same api as `Bloq.controlled` but as a step towards this direction; this PR ensures that we move away from using `cirq.Controlled` gate and always stay in Qualtran land. 

Note: This further regressed the CI time from ~17m to ~19m due to https://github.com/quantumlib/Qualtran/issues/793 and the fact that `Controlled` bloq is not a `GateWithRegisters`, it cannot use the underlying unitary even when it wraps a `GateWithRegisters` bloq, when doing cirq style simulations. Using the `Controlled` bloq in cirq-style simulations forces us to go via tensor contraction. Making it a `GateWithRegisters` would circumvent this but this can be addressed in follow-up PRs if needed. cc @mpharrigan

### CODE REVIEW

File: qualtran/_infra/gate_with_registers.py [hunk 1]
```diff
@@ -318,12 +319,42 @@ def controlled(
         control_qid_shape: Optional[Tuple[int, ...]] = None,
     ) -> 'cirq.Gate':
         # Multiple inheritance: use the `cirq.Gate` method.
-        return cirq.Gate.controlled(
+        from qualtran.cirq_interop import BloqAsCirqGate
```
Comment (mpharrigan) [inline qualtran/_infra/gate_with_registers.py:322]: update the comment above

File: qualtran/cirq_interop/_bloq_to_cirq_test.py [hunk 2]
```diff
@@ -160,75 +160,22 @@ def test_bloq_as_cirq_gate_left_register():
...
-                              │
-targets[3][0]: ───────────────×(y)────────────────────
-                              │
-targets[3][1]: ───────────────×(y)────────────────────
-                              │
-targets[3][2]: ───────────────×(y)────────────────────
-''',
-    )
+@attrs.frozen
+class TestSwapWithZero(Bloq):
```
Comment (tanujkhattar) [inline qualtran/cirq_interop/_bloq_to_cirq_test.py:164]: This is moot and I'll delete it.

File: qualtran/_infra/gate_with_registers.py [hunk 3]
```diff
@@ -318,12 +319,42 @@ def controlled(
...
+
+        controlled_gate = cirq.ControlledGate(
             self,
             num_controls=num_controls,
             control_values=control_values,
             control_qid_shape=control_qid_shape,
         )
+        ctrl_spec = CtrlSpec.from_cirq_cv(controlled_gate.control_values)
```
Comment (mpharrigan) [inline qualtran/_infra/gate_with_registers.py:330]: So we make the gate so we can get the corresponding `CtrlSpec` (and throw out the gate). Is this a kludge or is this fine? If kludge: open an issue to track simplifying

File: qualtran/cirq_interop/_bloq_to_cirq.py [hunk 4]
```diff
@@ -145,13 +146,15 @@ def _decompose_(self, qubits: Sequence[cirq.Qid]) -> cirq.OP_TREE:
         return self._decompose_with_context_(qubits)
 
     def _unitary_(self):
-        if self._num_qubits_() > 3:
-            # Prefer decomposition for large bloqs
-            return NotImplemented
```
Comment (tanujkhattar) [inline qualtran/cirq_interop/_bloq_to_cirq.py:150]: This had to be removed because it leads us to a situation where if we have a leaf bloq without a decomposition which acts on > 3 qubits but does override the `add_my_tensors` method; the cirq wrapper does not derive a unitary from it and raises an error which is not ideal

File: qualtran/_infra/gate_with_registers.py [hunk 5]
```diff
@@ -318,12 +319,42 @@ def controlled(
...
+        return BloqAsCirqGate(Controlled(self, ctrl_spec))
+
+    def _unitary_(self):
+        return NotImplemented
```
Comment (mpharrigan) [inline qualtran/_infra/gate_with_registers.py:334]: why this change? was this not the default?

File: qualtran/cirq_interop/cirq_interop.ipynb [hunk 6]
```diff
@@ -455,55 +455,6 @@
     "out_quregs"
    ]
   },
-  {
-   "cell_type": "markdown",
-   "id": "896e0392-3f1f-4965-bb13-0a541581ec25",
-   "metadata": {},
-   "source": [
-    "## Test `Cirq -> Bloqs -> Cirq` roundtrip using `QROM`\n",
```
Comment (mpharrigan) [inline qualtran/cirq_interop/cirq_interop.ipynb:463]: is this removal on purpose?

File: qualtran/_infra/gate_with_registers.py (hunk 3 above)
Comment (tanujkhattar) [reply qualtran/_infra/gate_with_registers.py:330]: Eventually, I'd like to change the `def controlled` method to use the bloq style `def controlled(ctrl_spec)` instead of cirq style `def controlled(self, num_controls, control_values, ...)` so this simplification wouldn't be required.

File: qualtran/bloqs/util_bloqs.py [hunk 7]
```diff
@@ -471,3 +471,41 @@ def on_classical_vals(self, reg: int) -> Dict[str, 'ClassicalValT']:
 
     def _t_complexity_(self) -> 'TComplexity':
         return TComplexity()
+
+
+@frozen
+class Power(Bloq):
```
Comment (mpharrigan) [inline qualtran/bloqs/util_bloqs.py:477]: is this addition on purpose? factor into its own PR? is it used elsewhere in the control stuff?

**Review by mpharrigan — APPROVED** [review]

File: qualtran/bloqs/util_bloqs.py (hunk 7 above)
Comment (tanujkhattar) [reply qualtran/bloqs/util_bloqs.py:477]: Yes, this addition is on purpose and was required to pass tests (eg: for phase estimation, where we use powers of controlled gate). I can factor into it's own PR but this PR will be blocked on the other PR being merged. I also need to add tests for this bloq

File: qualtran/cirq_interop/cirq_interop.ipynb (hunk 6 above)
Comment (tanujkhattar) [reply qualtran/cirq_interop/cirq_interop.ipynb:463]: Yeah, we've come a long way since this was written. The cirq interop probably needs to be updated now but maybe as a follow up. 


The idea is that now `BloqAsCirqGate` should be used to wrap only bloqs which are not already cirq gates (I've added an assertion to the constructor). `QROM` is a `GateWithRegisters` so should never be wrapped in a `BloqAsCirqGate`. This section was written when `QROM` was implemented in Cirq-FT and was not already a Bloq. 

Other examples of wrapping cirq-only gates in `CirqGateAsBloq` and bloq-only bloqs in `BloqAsCirqGate` are still part of the notebook.

File: qualtran/bloqs/util_bloqs.py (hunk 7 above)
Comment (mpharrigan) [reply qualtran/bloqs/util_bloqs.py:477]: no need to factor

File: qualtran/cirq_interop/_bloq_to_cirq.py (hunk 4 above)
Comment (mpharrigan) [reply qualtran/cirq_interop/_bloq_to_cirq.py:150]: sounds good
